fx-src/FirefoxStorage.js

```diff
@@ -0,0 +1,484 @@
+          this.cleanup();
+        }
+        throw e;
+      }
+    }
+    return this._dbconn;
+  }
+
+  cleanup() {
+    // Create backup file
+    const backupFile = this.file.leafName + ".corrupt";
+    Services.storage.backupDatabaseFile(this.file, backupFile);
+
+    if (this._dbconn) {
+      try {
+        // Honey-badger closing (we can't do anything about failure here)
+        this._dbconn.close();
+      } catch (e) {}
+    }
+    this.file.remove(false);
+    debug("all cleaned up");
+  }
+
+  requestAsyncClose() {
+    // ensure all statements are finalized
+    for (let statementString in this.statements) {
+      debug("finalizing "+statementString);
+      this.statements[statementString].finalize();
+    }
+
+    function connectionClosed() {
+      debug("connection closed");
+    }
+
+    // close the connection
+    if (this._dbconn) {
+      debug("closing the connection");
+      this._dbconn.asyncClose(connectionClosed());
+      this._dbconn = null;
+      this.statements = {};
+    }
+  }
+}
+
+export default class FirefoxAdapter extends BaseAdapter {
+  constructor(dbname) {
+    super();
+    this.operations = [];
+    this.busy = false;
+
+
+    // attempt creation - we don't need to use a transaction here since our
+    // operation queue means these will happen before other operations.
+    const statements = ["CREATE TABLE IF NOT EXISTS collection_metadata (collection_name TEXT PRIMARY KEY, last_modified INTEGER) WITHOUT ROWID;",
+          "CREATE TABLE IF NOT EXISTS collection_data (collection_name TEXT, record_id TEXT, record TEXT);",
+          "CREATE UNIQUE INDEX IF NOT EXISTS unique_collection_record ON collection_data(collection_name, record_id);"];
+
+    this.kintoStorage = new KintoStorage(dbname);
+
+    for (let stmt of statements) {
+      this.executeUpdate(stmt);
+    }
+  }
+
+  executeUpdate(sql) {
+    debug("requesting to execute statement: "+sql);
+    this.executeOperation(function(kintoStorage, complete) {
+      debug("executing statement: "+sql);
+      const statement = kintoStorage.getStatement(sql);
+
+      statement.executeAsync({
+        handleError: function(aError) {
+          debug("Error: " + aError.message);
+        },
+
+        handleCompletion: function(aReason) {
+          if (aReason != Ci.mozIStorageStatementCallback.REASON_FINISHED) {
+            debug("Query canceled or aborted!");
+          } else {
+            debug("Query complete");
+          }
+          complete();
+        }
+      });
+    });
+  }
+
+  executeOperation(op) {
+    debug("executeOperation");
+    if (!this.busy) {
+      debug("Executing operation now");
+      this.busy = true;
+      function executeNextOperation() {
+        debug("executeNextOperation");
+        const next = this.operations.shift();
+        if (next) {
+          debug("executing queued operation");
+          next(this.kintoStorage, executeNextOperation.bind(this));
```

Comment:
Can you explain why you're manually serializing operations here?

You should also think really hard about whether to use promises and Tasks, a la the second example here:

https://developer.mozilla.org/en-US/docs/Mozilla/JavaScript_code_modules/Sqlite.jsm#Open_Execute_Close